File: easy/20_Valid_Parentheses.cpp
ass Solution {
public:
    
    bool cmp(char f, char s) {
        
        if( (f  == '[') && (s == ']')) return true;
        if( (f  == '{') && (s == '}')) return true;
        if( (f  == '(') && (s == ')')) return true;
                           
        return false;
        
    }
    
    bool isValid(string s) {
        size_t sz = s.length();
        if((sz%2) == 1) return false;
        
        vector<char> cv;
        
        cv.push_back(s[0]);
        char c;
        
        for(int i = 1; i < s.length(); i++) {
         
         c = cv.back();
            
         if(cmp(c,s[i]) == true) {
             cv.pop_back();
         }else{
             cv.push_back(s[i]);
         }
        }
        
        if(cv.size() ==  0) return true;
        else return false;
        

    }
};

File: py/20-valid-parentheses.py
class Solution:
    
     def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """

        if len(s) %2 == 1:
            return False
        
        d = []
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '[' or s[i] == '{':
                d.append(s[i])
            if s[i] == ')':
                if len(d) is 0:
                    return False

                if d.pop() is not '(':
                    return False

            if s[i] == ']':
                if len(d) is 0:
                    return False

                if d.pop() is not '[':
                    return False

            if s[i] == '}':
                if len(d) is 0:
                    return False

                if d.pop() is not '{':
                    return False
        if len(d) is 0:
            return True
        else:
            return False

s = Solution()
print (s.isValid("[]()"))
print (s.isValid("["))
print (s.isValid("]["))
print (s.isValid("(("))
print (s.isValid("{[}]"))